packages/core/src/components/dialog/dialog.tsx

```diff
@@ -121,14 +126,19 @@ export class Dialog extends AbstractPureComponent2<DialogProps> {
             <Overlay {...this.props} className={Classes.OVERLAY_SCROLL_CONTAINER} hasBackdrop={true}>
                 <div className={Classes.DIALOG_CONTAINER} ref={this.props.containerRef}>
                     <div
-                        className={classNames(Classes.DIALOG, this.props.className)}
+                        className={classNames(Classes.DIALOG, this.props.footer != null && Classes.DIALOG_NO_PADDING, this.props.className)}
                         role="dialog"
                         aria-labelledby={this.props["aria-labelledby"] || (this.props.title ? this.titleId : undefined)}
                         aria-describedby={this.props["aria-describedby"]}
                         style={this.props.style}
                     >
                         {this.maybeRenderHeader()}
-                        {this.props.children}
+
+                        <div className={Classes.DIALOG_BODY_SCROLL_CONTAINER}>
+                            {this.props.children}
+                        </div>
+
+                        {this.maybeRenderFixedFooter()}
```

Comment:
The ergonomics / developer experience of this new footer feel a little awkward to me right now.

We've kept the Dialog component pretty simple so far, but if we're trying to add more functionality to improve & evolve its layout, we may want to reconsider the design choice of using only a single React component.

I think if we add 3 new components, we can create an API that's more legible while still preserving backwards-compatibility:

```ts
<Dialog isOpen={...} onClose={...}>
  <DialogHeader title="..." icon="..." isCloseButtonShown={...} />
  <DialogBody useOverflowScrollContainer={...}> // <-- scroll container would be enabled by default
    {/* children go here */}
  </DialogBody>
  <DialogFooter fixed={...} actions={...}>
    {/* footer left content goes here */}
  </DialogFooter>
</Dialog>
```

If you like this direction, I can help you implement these changes in this PR.

Reply:
I don't have a strong preference here. I've also seen the following pattern:

```
<Dialog
  header={IDialogHeader.basic(
    {
      title: "Title",
      icon: IconNames.FOLDER,
      isCloseButtonShown: false,
    })
  footer={IDialogFooter.fixed(
    {
      left: "",
      actions: ""}
    )}
  useOverflowScrollContainer={true}
>
  // Content
</Dialog>
```

That may help with discoverability of features? Not sure how back-compatible that would be.